src/Rules/Comparison/ImpossibleCheckTypeHelper.php

```diff
@@ -266,17 +289,40 @@ public function findSpecifiedType(
 					return null;
 				}
 			}
-			$types = TypeCombinator::union(
-				...array_column($sureNotTypes, 1),
-			);
```

Comment:
instead of union togehter all types within $sureNotTypes, we now take the $expr into account, and union only these refering to the same expr.